webapp/src/blocks.tsx

```diff
@@ -789,68 +802,472 @@ export class Editor extends srceditor.Editor {
+                                that.openExtension(name);
+                            }
+                        });
+                    }
+                })
+            }
+            return extraBlocks;
+        }
+
+        if (!snippets.isBuiltin(ns)) {
+            return getNonBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        } else {
+            return getBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        }
+    }
+
+    private flyoutBlockXmlCache: pxt.Map<Element[]> = {};
+    public showFlyout(treeRow: toolbox.ToolboxCategory) {
+        const { nameid, subns, icon, color, name, labelLineWidth } = treeRow;
+        const ns = nameid;
+
+        if (ns == 'search') {
+            this.showSearchFlyout();
+            return;
+        }
+
+        let xmlList: Element[] = [];
+        if (this.flyoutBlockXmlCache[ns + subns]) {
+            pxt.debug("showing flyout with blocks from flyout blocks xml cache");
+            xmlList = this.flyoutBlockXmlCache[ns + subns];
+            this.showFlyoutInternal_(xmlList);
+            return;
+        }
+
+        function createHeadingLabel() {
+            const categoryName = name ? name :
+                `${subns ? `${Util.capitalize(ns)} > ${Util.capitalize(subns)}` : Util.capitalize(ns)}`;
+            const iconClass = `blocklyTreeIcon${icon ? ns.toLowerCase() : 'Default'}`.replace(/\s/g, '');
+            let headingLabel = pxt.blocks.createFlyoutHeadingLabel(categoryName, color, icon, iconClass);
+            xmlList.push(headingLabel);
+        }
+        function createGroupLabel(group: string, groupicon?: string) {
+            let groupLabel = pxt.blocks.createFlyoutGroupLabel(pxt.Util.rlf(`{id:group}${group}`),
+                groupicon, labelLineWidth);
+            xmlList.push(groupLabel);
+        }
+
+        let that = this;
+        const filters = this.parent.state.editorState ? this.parent.state.editorState.filters : undefined;
+        const categoryState = filters ? (filters.namespaces && filters.namespaces[ns] != undefined ? filters.namespaces[ns] : filters.defaultState) : undefined;
+        function createBlocks(blocks: (toolbox.BlockDefinition | toolbox.ButtonDefinition)[]) {
+            blocks.sort((f1, f2) => {
+                // Sort the blocks
+                return (f2.attributes.weight != undefined ? f2.attributes.weight : 50)
+                    - (f1.attributes.weight != undefined ? f1.attributes.weight : 50);
+            }).forEach((block) => {
+                let blockXml: Element;
+                if (block.type == "button") {
+                    blockXml = that.getButtonXml(block as toolbox.ButtonDefinition);
+                } else {
+                    blockXml = that.getBlockXml(block as toolbox.BlockDefinition);
+                }
+                if (blockXml) xmlList.push(blockXml);
+            })
+        }
+
+        if (this.showFlyoutInternal(treeRow, createHeadingLabel, createGroupLabel, createBlocks)) {
```

Comment:
What is the difference between this method and the one with the underscore and can we eliminate one of them?

Reply:
showFlyoutInternal_ is used the one that calls Blockly and shows the blocks (takes in an xml list of blocks). 
showFlyoutInternal_ Calls blockly, defined in blocks.tsx
showFlyoutInternal in the superclass, shared between the two editors
showFlyout defined in each editor, defines the different helper methods needed ie: createHeading, createGroup, etc and calls showFlyoutInternal (in the super class).

Reply:
Renamed showFlyoutInternal to abstractShowFlyout perhaps to make it clearer. Let me know if that makes sense.